python/glow/gwas/log_reg.py

```diff
@@ -126,34 +126,91 @@ def _logistic_null_model_predictions(
 
     # Store 0 as prediction for samples with missing phenotypes
     remapped_predictions = np.zeros(y.shape)
-    remapped_predictions[y_mask] = predictions
+    remapped_predictions[mask] = predictions
     return remapped_predictions
 
 
-@dataclass
-class LogRegState:
-    inv_CtGammaC: NDArray[(Any, Any), Float]
-    gamma: NDArray[(Any, Any), Float]
-    Y_res: NDArray[(Any, Any), Float]
+def _prepare_one_phenotype(C: NDArray[(Any, Any), Float], row: pd.Series) -> pd.Series:
+    '''
+    Creates the broadcasted information for one (phenotype, offset) pair. The returned series
+    contains the information eventually stored in a LogRegState.
+
+    This function accepts and returns a pandas series for integration with Pandas UDFs and
+    pd.DataFrame.apply.
+    '''
+    y = row['values']
+    mask = ~np.isnan(y)
+    y_pred = _logistic_null_model_predictions(y, C, mask, row.get('offset'))
+    y_res = np.nan_to_num(y - y_pred)
+    gamma = y_pred * (1 - y_pred)
+    CtGammaC = C.T @ (gamma[:, None] * C)
+    inv_CtGammaC = np.linalg.inv(CtGammaC)
+    row.label = str(row.label)
+    row.drop(['values', 'offset'], inplace=True, errors='ignore')
```

Comment:
You mean to remove the `inplace`? I don't think it would make a significant perf difference, no. The whole null model fitting process doesn't actually matter much for perf since it is MUCH faster (like tens of seconds vs hours) than the association scan. It was only a problem when we were fitting all the models serially.

The main reason I used `row.drop` here was for simplicity -- otherwise we need separate cases for the loco vs non-loco paths.